### PR TITLE
[FixCode] Add @escaping when overriding mismatch is because of it.

### PR DESCRIPTION
**Pull request by nkcsgexi**
[FixCode] Add @escaping when overriding mismatch is because of it. rdar://27814862

```
With the change of default escaping behavior, users' existing code overriding
objc functions may need to add @escaping to make the overriding match as before. This
patch checks if an overriding mismatch is due to the lacking of @escaping and add
it as a fixit.
```

#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

…ar://27814862

With the change of default escaping behavior, users' existing code overriding
objc functions may need to add @escaping to make the overriding match as before. This
patch checks if an overriding mismatch is due to the lacking of @escaping and add
it as a fixit.

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci Please smoke test

**Comment by nkcsgexi** [comment]
@milseman @jrose-apple Could you please review this?

File: lib/Sema/MiscDiagnostics.cpp [hunk 1]
```diff
@@ -1184,6 +1184,39 @@ bool swift::fixItOverrideDeclarationTypes(TypeChecker &TC,
...
   };
 
+  // Check if overriding fails because we lack @escaping attribute on the function
+  // type repr.
+  auto checkTypeMissingEscaping = [&](Type overrideTy, Type baseTy,
+                                      SourceRange typeRange) -> bool {
+    // Fix-it needs position to apply.
+    if (typeRange.isInvalid())
+      return false;
+    auto overrideFnTy = overrideTy->getCanonicalType()->getAs<FunctionType>();
```
Comment (jrose-apple) [inline lib/Sema/MiscDiagnostics.cpp]: I think the lesson from the other patch was that we should *not* attempt to match types or ext-info; if we have no better reason why the override failed then being two AnyFunctionTypes mismatching in `@escaping` is a good enough reason for the diagnostic.

Comment (nkcsgexi) [inline lib/Sema/MiscDiagnostics.cpp]: But overriding match is strict. I'd like to have the fixit addresses the error for sure.

**Comment by nkcsgexi** [comment]
@swift-ci Please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci Please smoke test

File: lib/Sema/MiscDiagnostics.cpp (hunk 1 above)
Comment (jrose-apple) [inline lib/Sema/MiscDiagnostics.cpp]: Overriding isn't *this* strict. This doesn't catch
- covariant return types / contravariant parameters
- optionality changes for class types
- anything involving generics

(Also, the `getCanonicalType` isn't necessary; `getAs` looks through things.)

**Pull request merged**